compiler/rustc_hir_typeck/src/fn_ctxt/adjust_fulfillment_errors.rs

```diff
@@ -21,8 +21,15 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             return false;
         };
         let hir = self.tcx.hir();
-        let hir::Node::Expr(expr) = hir.get(hir_id) else {
-            return false;
+        let expr = match hir.get(hir_id) {
+            hir::Node::Expr(expr) => *expr,
+            hir::Node::Ty(hir::Ty { kind: hir::TyKind::Path(qpath), span, .. }) => {
+                // HACK: hir_id corresponds to a type in a binding. In order to reuse as much code
+                // as possible in the logic below, we "transmute" the hir::Ty into an analoguous
+                // hir::Expr. This let's us point to `T` in `let x: Type<T>;`.
+                hir::Expr { hir_id, kind: hir::ExprKind::Path(*qpath), span: *span }
+            }
+            _ => return false,
```

Comment:
You should avoid this hack by storing a node rather than an expr here

Reply:
I was trying to keep the diff shorter, but I can do that.